File: red/src/validate.ts
import { managed_errors, registry } from "colors-compute-red";
import { readFileSync } from "node:fs";
const managedProviders = JSON.parse(readFileSync(new URL("../resources/managed-providers.json", import.meta.resolve("colors-compute-red")), "utf8"));
export type Opts = Record<string, any>;
export const missing = (value: any) => value == null || typeof value === "string" && !value.trim();
export const registryEnabled = (opts: Opts) => !missing(opts["digitalocean-registry-tier"]);
export const registryName = (opts: Opts) => String(opts.profile ?? "").toLowerCase().replace(/[^a-z0-9-]/g, "");
export const computeRequest = (opts: Opts) => ({ legacy_state_keys: [`${opts.profile}/cluster.tfstate`] });
export const envErrors = (env: Opts) => env.COLORS_PAR_PROFILE ? ["COLORS_PAR_PROFILE is set; profile must come from colors.yml only"] : [];
export function stateErrors(opts: Opts): string[] {
  const errors = ["profile", "workdir", "provider-compute", "provider-backend", "compute-prevent-destroy"].filter(key => missing(opts[key])).map(key => `:${key} is required`);
  if (!missing(opts.profile) && !/^[A-Za-z0-9][A-Za-z0-9_-]{0,62}$/.test(String(opts.profile))) errors.push(":profile must be a safe identifier");
  const providers = Object.keys(managedProviders).sort();
  if (!providers.includes(opts["provider-compute"])) errors.push(":provider-compute must be one of " + providers.join(", "));
  if (typeof opts["compute-prevent-destroy"] !== "boolean") errors.push(":compute-prevent-destroy must be true or false");
  if (!errors.length) errors.push(...managed_errors(opts, computeRequest(opts)));
  if (registryEnabled(opts)) {
    if (opts["provider-compute"] !== "digitalocean") errors.push(":digitalocean-registry-tier requires :provider-compute digitalocean");
    if (!["starter", "basic", "professional"].includes(opts["digitalocean-registry-tier"])) errors.push(":digitalocean-registry-tier must be starter, basic, or professional");
    if (!/^[a-z0-9][a-z0-9-]{1,62}$/.test(registryName(opts))) errors.push("the profile-derived registry name is not a valid DigitalOcean registry name");
  }
  return errors;
}
export function secretErrors(opts: Opts, event: string): string[] {
  const entries = registry as Opts;
  const provider = entries.compute[opts["provider-compute"]]?.secrets ?? [];
  const backend = entries.backend[opts["provider-backend"]]?.secrets ?? [];
  const required: string[] = ["create", "delete"].includes(event) ? [...provider, ...backend] : event === "check" ? [...backend, ...(registryEnabled(opts) ? ["do-token"] : [])] : event === "kubeconfig" ? backend : event === "registry" ? ["do-token"] : [];
  return [...new Set(required)].filter(key => missing(opts[key])).map(key => "required credential is not set: COLORS_PAR_" + key.toUpperCase().replaceAll("-", "_"));
}

File: blue/src/package_doks_blue/validate.py
"""Desired-state and credential validation before workflow effects."""
import re
from colors_compute import registry
from colors_compute.managed import managed_errors, _recipes


def missing(value):
    return value is None or isinstance(value, str) and not value.strip()


def registry_enabled(opts):
    return not missing(opts.get('digitalocean-registry-tier'))


def registry_name(opts):
    return re.sub('[^a-z0-9-]', '', str(opts.get('profile', '')).lower())


def compute_request(opts):
    return {'legacy_state_keys': [f"{opts['profile']}/cluster.tfstate"]}


def env_errors(env):
    return ['COLORS_PAR_PROFILE is set; profile must come from colors.yml only'] if env.get('COLORS_PAR_PROFILE') else []


def state_errors(opts):
    errors = [f':{k} is required' for k in ('profile', 'workdir', 'provider-compute', 'provider-backend', 'compute-prevent-destroy') if missing(opts.get(k))]
    if not missing(opts.get('profile')) and not re.fullmatch('[A-Za-z0-9][A-Za-z0-9_-]{0,62}', str(opts['profile'])):
        errors.append(':profile must be a safe identifier')
    providers = sorted(_recipes())
    if opts.get('provider-compute') not in providers:
        errors.append(':provider-compute must be one of ' + ', '.join(providers))
    if not isinstance(opts.get('compute-prevent-destroy'), bool):
        errors.append(':compute-prevent-destroy must be true or false')
    if not errors:
        errors.extend(managed_errors(opts, compute_request(opts)))
    if registry_enabled(opts):
        if opts.get('provider-compute') != 'digitalocean':
            errors.append(':digitalocean-registry-tier requires :provider-compute digitalocean')
        if opts.get('digitalocean-registry-tier') not in ('starter', 'basic', 'professional'):
            errors.append(':digitalocean-registry-tier must be starter, basic, or professional')
        if not re.fullmatch('[a-z0-9][a-z0-9-]{1,62}', registry_name(opts)):
            errors.append('the profile-derived registry name is not a valid DigitalOcean registry name')
    return errors


def secret_errors(opts, event):
    provider = registry()['compute'].get(opts.get('provider-compute'), {}).get('secrets', [])
    backend = registry()['backend'].get(opts.get('provider-backend'), {}).get('secrets', [])
    required = provider + backend if event in ('create', 'delete') else backend + (['do-token'] if registry_enabled(opts) else []) if event == 'check' else backend if event == 'kubeconfig' else ['do-token'] if event == 'registry' else []
    return ['required credential is not set: COLORS_PAR_' + key.upper().replace('-', '_') for key in dict.fromkeys(required) if missing(opts.get(key))]
